File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/testFixtures/java/io/airbyte/cdk/integrations/standardtest/destination/JdbcDestinationAcceptanceTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.standardtest.destination;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.Arrays;
import java.util.Optional;
import java.util.function.Function;
import org.jooq.Record;

public abstract class JdbcDestinationAcceptanceTest extends DestinationAcceptanceTest {

  protected final ObjectMapper mapper = new ObjectMapper();

  protected JsonNode getJsonFromRecord(final Record record) {
    return getJsonFromRecord(record, x -> Optional.empty());
  }

  protected JsonNode getJsonFromRecord(final Record record, final Function<Object, Optional<String>> valueParser) {
    final ObjectNode node = mapper.createObjectNode();

    Arrays.stream(record.fields()).forEach(field -> {
      final var value = record.get(field);

      final Optional<String> parsedValue = valueParser.apply(value);
      if (parsedValue.isPresent()) {
        node.put(field.getName(), parsedValue.get());
      } else {
        switch (field.getDataType().getTypeName()) {
          case "varchar", "nvarchar", "jsonb", "json", "other":
            final var stringValue = (value != null ? value.toString() : null);
            DestinationAcceptanceTestUtils.putStringIntoJson(stringValue, field.getName(), node);
            break;
          default:
            node.put(field.getName(), (value != null ? value.toString() : null));
        }
      }
    });
    return node;
  }

}

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/testFixtures/kotlin/io/airbyte/cdk/integrations/standardtest/destination/JdbcDestinationAcceptanceTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.standardtest.destination

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import java.util.*
import java.util.function.Function
import org.jooq.Field
import org.jooq.Record

abstract class JdbcDestinationAcceptanceTest : DestinationAcceptanceTest() {
    protected val mapper: ObjectMapper = ObjectMapper()

    protected fun getJsonFromRecord(record: Record): JsonNode {
        return getJsonFromRecord(record, Function { x: Any? -> Optional.empty() })
    }

    protected fun getJsonFromRecord(
        record: Record,
        valueParser: Function<Any, Optional<String>>
    ): JsonNode {
        val node = mapper.createObjectNode()

        Arrays.stream(record.fields()).forEach { field: Field<*> ->
            val value = record[field]
            val parsedValue = valueParser.apply(value)
            if (parsedValue.isPresent) {
                node.put(field.name, parsedValue.get())
            } else {
                when (field.dataType.typeName) {
                    "varchar",
                    "nvarchar",
                    "jsonb",
                    "json",
                    "other" -> {
                        val stringValue = (value?.toString())
                        DestinationAcceptanceTestUtils.putStringIntoJson(
                            stringValue,
                            field.name,
                            node
                        )
                    }
                    else -> node.put(field.name, (value?.toString()))
                }
            }
        }
        return node
    }
}
